### PR TITLE
:recycle: refactor indexing + :scissors: decouple stride & window + :sparkles: support segment idxs

### PR DESCRIPTION
**Pull request by jvdd**
## :recycle: Refactor indexing

* :bug: fix bug with `vectorized=True` for strided rolling
  - [x] vectorized support for single feature windows
  - [x] vectorized support for empty feature windows
  - [x] test the above
* :recycle: refactor the strided window segmentation (& indexing)
  - [x] add `include_final_window` argument to FeatureCollection `.calculate` (& StridedRolling)
  - [x] update docs
  - [x] test the above
  - [ ] remove support for TimeSequenceStridedRolling 
    *=> Decided to not do this in this PR. We will leave this for another PR.*
* :see_no_evil: undo breaking change from #62 (to make code backwards compatible)
  - [x] revert the default `window_idx` argument in `FeatureCollection.calculate` to `"end"` 
* :robot: extend test matrix & update dependencies
  - [x] Add Python 3.10 to the test matrix
  - [x] Update dependencies (& remove locked statsmodels dependency - https://github.com/blue-yonder/tsfresh/issues/897)

## :scissors: Decouple stride
We (@jonasvdd, @emield12, and @jvdd) believe that the stride should not be hardly coupled with a `FeatureDescriptor`. Therefore, to make tsflex more flexible (:wink:) we make the stride argument optional for `FeatureDescriptor` and `MultipleFeatureDescriptor` and add the functionality to pass your stride(s) to the `FeatureCollection.calculate` method. 

* :eyes: externally visible changes
  - [x] make stride optional in `FeatureDescriptor`
  - [x] add stride argument to `FeatureCollection.calculate`
  - [x] `FeatureDescriptor` / `FeatureCollection.calculate` should accept multiple strides
    - [x] `StridedRolling` should accept multiple strides
    - [x] test the above 
  - [x] remove stride from output column name
      - [x] update reduce method
      - [x] update reduce method tests
  - [x] update the logging to handle multiple strides
      - [x] extend logging tests
* :package: internal changes
  - [x] change `stride` -> `strides`: which is either a list of stride sizes (float or pd.Timedelta) or None (in `StridedRolling` and `StridedRollingFactory`)
  - [x] identify feature descriptors (FD) based on their window - output names
      - [x] set interection after `StridedRolling` search sorted
              TODO: moet geoptimaliseerd worden

## :sparkles: Support setpoints

- [x] support setpoints
- [x] test the above
  => *Note: we allow setpoints of different timezones as the `np.datetime64` conversion of these allow comparison..*
- [ ] trim range if not in data
  => *Decided to not do this! (as this is somewhat an ambiguous operation)
  As using segment indexes is already an advanced operation, it is the user its responsability to either trim the segmented indexes or [make their features robust](https://predict-idlab.github.io/tsflex/features/utils.html#tsflex.features.utils.make_robust)*.

## :see_no_evil: other stuff
* :bug: fix bug with `features.logger` that does not handle numeric window & strides
  - [x] improve parsing for window and stride values in `_parse_logging_execution_to_df`
  - [x] test the above 
* :eyes: other minor stuff
  - [x] support offline data load
  - [x] warn the user with a `RuntimeWarining` when the data its index (passed to `FeatureCollection.calculate`) is not monotonically increasing 
  - [x] test the above

### CODE REVIEW

File: tests/test_features_feature_collection.py
Comment (jonasvdd) [inline tests/test_features_feature_collection.py:974]: Why is the `output_names` = 'max_` when the vectorized function is `np.min`?

**Review by jonasvdd — APPROVED** [review]
Looks good to me, just need to think sometime later about the `TimeSequenceStridedRolling`

File: tsflex/features/feature_collection.py
Comment (jvdd) [reply tsflex/features/feature_collection.py]: Was showing the weird sentence completion functionality of copilot to some colleagues :see_no_evil:

**Comment by jvdd** [comment]
FYI: all tests that fail now are due to `TimeIndexSampleStridedRolling`, which I did not update accordingly in this PR => I believe we should decide on whether we include the support for this or drop it... @jonasvdd @emield12

File: tests/test_strided_rolling.py
Comment (jvdd) [inline tests/test_strided_rolling.py:174]: Do you agree with this behavior?

File: tests/test_strided_rolling.py
Comment (jvdd) [inline tests/test_strided_rolling.py:186]: Do you agree with this behavior @jonasvdd? Now only 1 incomplete window (i.e., < window size) is included because of the flag

File: tsflex/features/segmenter/strided_rolling.py
Comment (jvdd) [inline tsflex/features/segmenter/strided_rolling.py:173]: @jonasvdd guess we need to have a short discussion about this

Comment (jvdd) [reply tsflex/features/segmenter/strided_rolling.py:173]: Working on it :runner:

**Comment by jvdd** [comment]
macOS-3.10 test is failing due to problems with installing matrix-profile https://github.com/blue-yonder/tsfresh/issues/937

**Comment by jonasvdd** [comment]
is this ready for a re-review? @jvdd

**Comment by arturdaraujo** [comment]
Waiting for this awesome PR to be merged! A lot of cool stuff!

**Review by jonasvdd — APPROVED** [review]

File: tsflex/features/feature.py
Comment (jonasvdd) [reply tsflex/features/feature.py]: This is resolved in the current PR!

File: tsflex/features/feature_collection.py
Comment (jonasvdd) [reply tsflex/features/feature_collection.py]: removed

File: tsflex/features/feature_collection.py
Comment (jonasvdd) [reply tsflex/features/feature_collection.py]: Altered the message:
```python
            "When using `segment_XXX_idxs`; each output name - series_input combination"
            + " can only have 1 window (or None)"
```

File: tsflex/features/feature.py
Comment (jonasvdd) [reply tsflex/features/feature.py]: Discussed with @jvdd, this is kept for backward compatibility

File: tsflex/features/feature.py
Comment (jonasvdd) [reply tsflex/features/feature.py:240]: Discussed with @jvdd, this is kept for backward compatibility

File: tsflex/utils/data.py
Comment (jonasvdd) [reply tsflex/utils/data.py]: But might be out of scope for this PR

Comment (jvdd) [reply tsflex/utils/data.py]: Seems to work fine :+1:

File: tsflex/features/segmenter/strided_rolling.py
Comment (jvdd) [reply tsflex/features/segmenter/strided_rolling.py]: Will do that! We'll make it the user its responsability to properly trim the segment indexes.

File: tsflex/features/segmenter/strided_rolling.py
Comment (jvdd) [reply tsflex/features/segmenter/strided_rolling.py:500]: Agree! I guess I skipped this bc it is only useful for the `TimeIndexSampleStridedRolling`

File: tsflex/features/segmenter/strided_rolling.py
Comment (jvdd) [reply tsflex/features/segmenter/strided_rolling.py]: Convert to UTC (allowing comparison with the segment_idxs).
Added this as a comment as well!

Comment (jonasvdd) [reply tsflex/features/segmenter/strided_rolling.py]: sounds perfect!

**Comment by jvdd** [comment]
Tests on windows are failing bc poetry installation issue https://github.com/snok/install-poetry/issues/94

**Pull request merged**
